### PR TITLE
add more examples

### PR DESCRIPTION
**Pull request by kira-lin**
This PR intends to add more examples to show how raydp can work with other libraries. 
You are welcome to leave a comment, if you want to see an example on a certain workload.
- [x] XGBoost
- [x] RaySGD
- [x] Horovod

### CODE REVIEW

File: examples/data_process.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            .drop("dropoff_latitude") \
+            .drop("passenger_count") \
+            .drop("key")
+    return data
+
+def nyc_taxi_preprocess(data):
+    data = clean_up(data)
+    data = add_time_features(data)
+    data = add_distance_features(data)
+    return drop_col(data)
```
Comment (ConeyLiu) [inline examples/data_process.py:91]: add one blank line in the end.

File: examples/horovod_nyctaxi.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    # Start horovod workers on Ray
+    from horovod.ray import RayExecutor
+    settings = RayExecutor.create_settings(500)
+    executor = RayExecutor(settings, num_hosts=1, num_slots=2, cpus_per_slot=2)
+    executor.start()
+    # torch_ds, num_features = executor.execute_single(process_data)
+    torch_ds, num_features = process_data()
+    executor.run(train_fn, args=[torch_ds, num_features])
+    raydp.stop_spark()
+    ray.shutdown()
```
Comment (ConeyLiu) [inline examples/horovod_nyctaxi.py:135]: same as above

File: examples/pytorch_nyctaxi.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+# Create a distributed estimator based on the raydp api
+estimator = TorchEstimator(num_workers=4, model=nyc_model, optimizer=optimizer, loss=criterion,
+                           feature_columns=features, label_column="fare_amount", batch_size=256,
+                           num_epochs=30)
+# Train the model
+estimator.fit_on_spark(train_df, test_df)
+# shutdown raydp and ray
+estimator.shutdown()
+raydp.stop_spark()
+ray.shutdown()
```
Comment (ConeyLiu) [inline examples/pytorch_nyctaxi.py:79]: same as above

**Comment by ConeyLiu** [comment]
Hi @kira-lin, very perfect, thanks for the awesome update. Could we also test those examples with CI too?

**Comment by kira-lin** [comment]
> Hi @kira-lin, very perfect, thanks for the awesome update. Could we also test those examples with CI too?

Sure, I have updated the CI. But will github run the workflow in this commit?

File: .github/workflows/raydp.yml [hunk 4]
```diff
@@ -59,3 +65,13 @@ jobs:
...
+      - name: Test Examples
+        run: |
+          ray start --head
+          python examples/random_nyctaxi.py
+          python examples/pytorch_nyctaxi.py
+          python examples/tensorflow_nyctaxi.py
+          python examples/horovod_nyctaxi.py
+          python examples/xgboost_ray_nyctaxi.py
+          python examples/raysgd_nyctaxi.py
+          ray stop
```
Comment (ConeyLiu) [inline .github/workflows/raydp.yml:77]: code style, same as the following.

**Comment by ConeyLiu** [comment]
Hi @kira-lin, there are some other `jupyter` examples, do you plan to convert them as well?

**Comment by kira-lin** [comment]
> Hi @kira-lin, there are some other `jupyter` examples, do you plan to convert them as well?

I think we should keep the DLRM notebook, because it's nice to have visualization in jupyter when training such large model.

For the Titanic, we already have a tensorflow nyctaxi example to shownhow to use the tensorflow estimator. And another random dataset generator is needed in case we want to run it in CI. So I think we can just keep it.

File: .gitignore [hunk 5]
```diff
@@ -8,13 +8,14 @@ dist/
...
 .DS_Store
 
 .vscode
-
 examples/.ipynb_checkpoints/
 
 # Vim temp files
 *.swp
 *.swo
+*.csv
```
Comment (ConeyLiu) [inline .gitignore:21]:
```suggestion
*.csv

```

File: examples/data_process.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
```
Comment (ConeyLiu) [inline examples/data_process.py:95]:
```suggestion

```

File: examples/horovod_nyctaxi.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    from horovod.ray import RayExecutor
+    settings = RayExecutor.create_settings(500)
+    executor = RayExecutor(settings, num_hosts=1, num_slots=1, cpus_per_slot=1)
+    executor.start()
+    # torch_ds, num_features = executor.execute_single(process_data)
+    torch_ds, num_features = process_data()
+    executor.run(train_fn, args=[torch_ds, num_features])
+    raydp.stop_spark()
+    ray.shutdown()
+
```
Comment (ConeyLiu) [inline examples/horovod_nyctaxi.py:136]:
```suggestion

```

File: examples/random_nyctaxi.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    'fare_amount':fare_amount,
+    'pickup_datetime': date,
+    'pickup_longitude': pick_long,
+    'pickup_latitude': pick_lat,
+    'dropoff_longitude': drop_long,
+    'dropoff_latitude': drop_lat,
+    'passenger_count': passenger_count
+    })
+csv_path = os.path.dirname(os.path.realpath(__file__)) + '/fake_nyctaxi.csv'
+df.to_csv(csv_path, index=False)
```
Comment (ConeyLiu) [inline examples/random_nyctaxi.py:41]:
```suggestion
df.to_csv(csv_path, index=False)

```

**Review by ConeyLiu — COMMENTED** [review]
code style

**Comment by kira-lin** [comment]
I do not know exactly why but CI seems to be fixed. Here is 2 types of errors I found in previous runs:
1. Pytorch complains about the shape. Something like:
`ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 256])`
Once I just rerun the workflow, and it's solved. I guess it has something to do with pytorch version, like some mirror has stale version, but I'm not sure. I don't have this problem on my machine with either pytorch 1.8.1 or 1.7.1.
2. Workflow sometimes hangs because ray actor cannot be scheduled due to lack of resource. It seems like some tasks are not releasing resources, but I didn't found any. It never happened on my machine, even when I run them in a 2-cpu ray cluster just like the vm used by github. And now it's passing without significant changes to code. I guess it's because github workflow machines are too over-used. 

@ConeyLiu Please review again.

**Comment by ConeyLiu** [comment]
thanks @kira-lin, LGTM

**Comment by ConeyLiu** [comment]
Hi @kira-lin, it breaks the CI. Please refer to: https://discuss.pytorch.org/t/error-expected-more-than-1-value-per-channel-when-training/26274/4.

**Pull request merged**